.ci/upgrade_nodes_ci.sh

```diff
@@ -0,0 +1,595 @@
+          fi
+          if (( cv != last_cv )); then
+            log "ℹ️ Consensus version changed from $last_cv to $cv at height $h"
+          fi
+          last_cv="$cv"
+          last_h="$h"
+        fi
+      else
+        log "WARN: invalid consensus or height: cv='$cv' h='$h'"
+      fi
+      sleep "$interval"
+      elapsed=$((elapsed + interval))
+    done
+    echo "❌ Timed out waiting for stable consensus version" >&2
+    return 1
+  fi
+}
+
+wait_for_height_increase_window() {
+  local previous_height="$1"
+  local duration="${2:-60}"
+  local interval="${3:-5}"
+
+  local elapsed=0
+  local increased=0
+  local current_height="$previous_height"
+
+  log "Waiting ${duration}s window to see height increase above $previous_height…"
+
+  while (( elapsed < duration )); do
+    if current_height="$(get_latest_height)"; then
+      log "Current height=${current_height}"
+      if (( current_height > previous_height )); then
+        increased=1
+      fi
+    else
+      log "WARN: could not fetch latest height"
+    fi
+    sleep "$interval"
+    elapsed=$((elapsed + interval))
+  done
+
+  if (( increased == 1 )); then
+    log "✅ Height increased from $previous_height to $current_height during upgrade window."
+    return 0
+  else
+    echo "❌ Timeout: height did not increase above $previous_height during ${duration}s." >&2
+    return 1
+  fi
+}
+
+########################################
+# MAIN
+########################################
+
+log "Starting upgrade_nodes_ci.sh"
+log "total_validators=${total_validators}, total_clients=${total_clients}, network_id=${network_id}"
+log "EXPECTED_MAX_CONSENSUS_VERSION=${EXPECTED_MAX_CONSENSUS_VERSION:-<none>} WAIT_BETWEEN_UPGRADES=${WAIT_BETWEEN_UPGRADES}"
+
+# 1. Build release snarkos once (without heights override yet).
+download_and_build_latest_snarkos 0
+
+# 2–4. Probe versions, compute heights, rebuild release & PR with those heights.
+derive_consensus_env_from_version
+
+# From here on, we use:
+#   - SNARKOS_RELEASE_BIN  => rebuilt with release heights
+#   - snarkos (PR)         => rebuilt with PR heights
+
+log "Cleaning dev stores with release binary..."
+for ((node_index = 0; node_index < total_validators + total_clients; node_index++)); do
+  "$SNARKOS_RELEASE_BIN" clean "--dev=$node_index" "--network=$network_id"
+done
+
+for ((validator_index = 0; validator_index < total_validators; validator_index++)); do
+  log_file="$log_dir/validator-$validator_index.log"
+  start_node "$SNARKOS_RELEASE_BIN" "$validator_index" "validator" "$log_file"
+  sleep 1
+done
+
+for ((client_index = 0; client_index < total_clients; client_index++)); do
+  node_index=$((client_index + total_validators))
+  log_file="$log_dir/client-$client_index.log"
+  start_node "$SNARKOS_RELEASE_BIN" "$node_index" "client" "$log_file"
+  if (( client_index < total_clients-1 )); then
+    sleep 1
+  fi
+done
+
+wait_for_nodes "$total_validators" "$total_clients"
+wait_for_highest_consensus_version 900 10
```

Comment:
We are calling `wait_for_highest_consensus_version` both before and after the 1-by-1 upgrades. 

Is `EXPECTED_MAX_CONSENSUS_VERSION` set in both cases? Is it set to the same value? Should it be?